File: src/dev_health_ops/api/auth/routers/invites.py
from __future__ import annotations

import logging
import uuid as uuid_mod
from datetime import datetime, timezone
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Request
from pydantic import BaseModel
from sqlalchemy import select

from dev_health_ops.api.services.auth import AuthenticatedUser
from dev_health_ops.api.services.invites import (
    accept_invite as accept_org_invite,
)
from dev_health_ops.api.services.invites import (
    validate_invite as validate_org_invite,
)
from dev_health_ops.api.utils.audit import emit_audit_log
from dev_health_ops.api.utils.errors import error_detail

from dev_health_ops.models.audit import AuditAction, AuditResourceType
from dev_health_ops.models.users import Membership, Organization, User

from .common import _issue_membership_tokens, _require_uuid, _slugify_org_name
from .dependencies import get_current_user

logger = logging.getLogger(__name__)

router = APIRouter()


class AcceptInviteRequest(BaseModel):
    token: str


class AcceptInviteResponse(BaseModel):
    access_token: str
    refresh_token: str
    token_type: str = "bearer"
    expires_in: int
    org_id: str
    org_name: str
    role: str


class OnboardRequest(BaseModel):
    action: str
    org_name: str | None = None
    invite_code: str | None = None


class OnboardResponse(BaseModel):
    access_token: str
    refresh_token: str
    token_type: str = "bearer"
    expires_in: int
    org_id: str
    org_name: str
    role: str


@router.post("/accept-invite", response_model=AcceptInviteResponse)
async def accept_invite(
    payload: AcceptInviteRequest,
    user: Annotated[AuthenticatedUser, Depends(get_current_user)],
    request: Request,
) -> AcceptInviteResponse:
    from dev_health_ops.api.auth.router import get_postgres_session

    async with get_postgres_session() as db:
        try:
            user_uuid = uuid_mod.UUID(user.user_id)
        except ValueError as exc:
            raise HTTPException(
                status_code=401,
                detail=error_detail("Invalid user identity"),
            ) from exc

        user_result = await db.execute(select(User).where(User.id == user_uuid))
        db_user = user_result.scalar_one_or_none()
        if db_user is None:
            raise HTTPException(
                status_code=401,
                detail=error_detail("User not found"),
            )

        invite = await validate_org_invite(db, payload.token)
        if invite is None:
            raise HTTPException(
                status_code=400,
                detail=error_detail("Invalid or expired invite"),
            )

        org_result = await db.execute(
            select(Organization).where(Organization.id == invite.org_id)
        )
        org = org_result.scalar_one_or_none()
        if org is None:
            raise HTTPException(
                status_code=404,
                detail=error_detail("Organization not found"),
            )

        try:
            db_user_id = _require_uuid(db_user.id, "db_user.id")
            membership = await accept_org_invite(db, invite, db_user_id)
        except ValueError as exc:
            raise HTTPException(
                status_code=400,
                detail=error_detail(str(exc)),
            ) from exc

        invite_org_id = _require_uuid(invite.org_id, "invite.org_id")
        membership_id = _require_uuid(membership.id, "membership.id")
        db_user_id = _require_uuid(db_user.id, "db_user.id")
        emit_audit_log(
            db,
            org_id=invite_org_id,
            action=AuditAction.MEMBER_JOINED,
            resource_type=AuditResourceType.MEMBERSHIP,
            resource_id=str(membership_id),
            user_id=db_user_id,
            description="Invite accepted",
            changes={
                "invite_id": str(invite.id),
                "user_id": str(db_user_id),
                "role": membership.role,
            },
            request=request,
        )

        await db.commit()
        token_pair = await _issue_membership_tokens(db, request, db_user, membership)

        return AcceptInviteResponse(
            access_token=token_pair.access_token,
            refresh_token=token_pair.refresh_token,
            token_type=token_pair.token_type,
            expires_in=token_pair.expires_in,
            org_id=str(org.id),
            org_name=str(org.name),
            role=str(membership.role),
        )


@router.post("/onboard", response_model=OnboardResponse)
async def onboard(
    payload: OnboardRequest,
    user: Annotated[AuthenticatedUser, Depends(get_current_user)],
    request: Request,
) -> OnboardResponse:
    from dev_health_ops.api.auth.router import get_postgres_session

    async with get_postgres_session() as db:
        try:
            user_uuid = uuid_mod.UUID(user.user_id)
        except ValueError as exc:
            raise HTTPException(
                status_code=401,
                detail=error_detail("Invalid user identity"),
            ) from exc

        user_result = await db.execute(select(User).where(User.id == user_uuid))
        db_user = user_result.scalar_one_or_none()
        if not db_user:
            raise HTTPException(
                status_code=401,
                detail=error_detail("User not found"),
            )

        membership_result = await db.execute(
            select(Membership.id).where(Membership.user_id == db_user.id)
        )
        if membership_result.first() is not None:
            raise HTTPException(
                status_code=400,
                detail=error_detail("Already onboarded"),
            )

        if payload.action == "join_org":
            if not payload.invite_code:
                raise HTTPException(
                    status_code=400,
                    detail=error_detail("invite_code is required"),
                )

            invite = await validate_org_invite(db, payload.invite_code)
            if invite is None:
                raise HTTPException(
                    status_code=400,
                    detail=error_detail("Invalid or expired invite"),
                )

            org_result = await db.execute(
                select(Organization).where(Organization.id == invite.org_id)
            )
            org = org_result.scalar_one_or_none()
            if org is None:
                raise HTTPException(
                    status_code=404,
                    detail=error_detail("Organization not found"),
                )

            try:
                db_user_id = _require_uuid(db_user.id, "db_user.id")
                membership = await accept_org_invite(db, invite, db_user_id)
            except ValueError as exc:
                raise HTTPException(
                    status_code=400,
                    detail=error_detail(str(exc)),
                ) from exc

            org_id = _require_uuid(org.id, "org.id")
            membership_id = _require_uuid(membership.id, "membership.id")
            db_user_id = _require_uuid(db_user.id, "db_user.id")
            emit_audit_log(
                db,
                org_id=org_id,
                action=AuditAction.MEMBER_JOINED,
                resource_type=AuditResourceType.MEMBERSHIP,
                resource_id=str(membership_id),
                user_id=db_user_id,
                description="Invite accepted during onboarding",
                changes={
                    "invite_id": str(invite.id),
                    "user_id": str(db_user_id),
                    "role": membership.role,
                },
                request=request,
            )

            await db.commit()
            token_pair = await _issue_membership_tokens(
                db, request, db_user, membership
            )

            return OnboardResponse(
                access_token=token_pair.access_token,
                refresh_token=token_pair.refresh_token,
                token_type=token_pair.token_type,
                expires_in=token_pair.expires_in,
                org_id=str(org.id),
                org_name=str(org.name),
                role=str(membership.role),
            )

        if payload.action != "create_org":
            raise HTTPException(
                status_code=400,
                detail=error_detail("Invalid action. Use 'create_org' or 'join_org'"),
            )

        org_name = payload.org_name or "My Organization"
        org_slug = f"{_slugify_org_name(org_name)}-{str(db_user.id)[:8]}"

        org = Organization(
            slug=org_slug,
            name=org_name,
            tier="community",
            is_active=True,
        )
        db.add(org)
        await db.flush()
        db_user_id = _require_uuid(db_user.id, "db_user.id")
        org_id = _require_uuid(org.id, "org.id")

        membership = Membership(
            user_id=db_user_id,
            org_id=org_id,
            role="owner",
            joined_at=datetime.now(timezone.utc),
        )
        db.add(membership)

        emit_audit_log(
            db,
            org_id=org_id,
            action=AuditAction.CREATE,
            resource_type=AuditResourceType.ORGANIZATION,
            resource_id=str(org_id),
            user_id=db_user_id,
            description="Organization created during onboarding",
            changes={"organization_name": org_name},
            request=request,
        )

        await db.commit()

        token_pair = await _issue_membership_tokens(db, request, db_user, membership)

        return OnboardResponse(
            access_token=token_pair.access_token,
            refresh_token=token_pair.refresh_token,
            token_type=token_pair.token_type,
            expires_in=token_pair.expires_in,
            org_id=str(org.id),
            org_name=str(org.name),
            role="owner",
        )

File: internal/api/session/invites.go
package session

import (
	"context"
	"errors"
	"net/http"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"

	"github.com/full-chaos/dev-health-ops/internal/api/audit"
	"github.com/full-chaos/dev-health-ops/internal/api/policy"
	"github.com/full-chaos/dev-health-ops/internal/api/pybody"
	"github.com/full-chaos/dev-health-ops/internal/api/pyjson"
	"github.com/full-chaos/dev-health-ops/internal/auth/signedtoken"
	"github.com/full-chaos/dev-health-ops/internal/pythonparity"
)

// errInviteRefused carries services/invites.py accept_invite's ValueError
// text, which the routes answer as a 400.
type errInviteRefused struct{ message string }

func (e errInviteRefused) Error() string { return e.message }

// pendingInvite is the org_invites columns accept_invite reads.
type pendingInvite struct {
	ID          uuid.UUID
	OrgID       uuid.UUID
	Role        string
	InvitedByID *uuid.UUID
	ExpiresAt   time.Time
}

// validateInvite is services/invites.py validate_invite: the pending,
// unexpired invite token names, nil when Python returns None. An expired
// invite is marked expired there, but the route then refuses the request
// and the session rolls back, so the mark is never written.
func (h handlers) validateInvite(ctx context.Context, tx pgx.Tx, token string, now time.Time) (*pendingInvite, error) {
	valid, err := signedtoken.Valid(token, h.Mail.Secret())
	if err != nil || !valid {
		return nil, err
	}
	var invite pendingInvite
	var status string
	err = tx.QueryRow(ctx, `SELECT id, org_id, role, invited_by_id, status, expires_at FROM org_invites WHERE token_hash = $1::text`,
		signedtoken.Hash(token)).Scan(&invite.ID, &invite.OrgID, &invite.Role, &invite.InvitedByID, &status, &invite.ExpiresAt)
	if errors.Is(err, pgx.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	if status != "pending" || invite.ExpiresAt.Before(now) {
		return nil, nil
	}
	return &invite, nil
}

// acceptInvite is services/invites.py accept_invite (its status and expiry
// re-check cannot fail after validateInvite in the same transaction): the
// new membership, or errInviteRefused when the user already belongs to
// the organization.
func (h handlers) acceptInvite(ctx context.Context, tx pgx.Tx, invite *pendingInvite, userID uuid.UUID, now time.Time) (uuid.UUID, error) {
	var member bool
	if err := tx.QueryRow(ctx, `SELECT EXISTS (SELECT 1 FROM memberships WHERE org_id = $1::uuid AND user_id = $2::uuid)`,
		invite.OrgID, userID).Scan(&member); err != nil {
		return uuid.UUID{}, err
	}
	if member {
		return uuid.UUID{}, errInviteRefused{"User is already a member of this organization"}
	}
	id := h.NewUUID()
	if err := insertMembership(ctx, tx, id, userID, invite.OrgID, invite.Role, invite.InvitedByID, now); err != nil {
		return uuid.UUID{}, err
	}
	_, err := tx.Exec(ctx, `UPDATE org_invites SET status = 'accepted', accepted_at = $2, updated_at = $2 WHERE id = $1::uuid`,
		invite.ID, now.UTC())
	return id, err
}

// organizationName reads organizations.name; found is false without a row.
func organizationName(ctx context.Context, tx pgx.Tx, id uuid.UUID) (name string, found bool, err error) {
	err = tx.QueryRow(ctx, `SELECT name FROM organizations WHERE id = $1::uuid`, id).Scan(&name)
	if errors.Is(err, pgx.ErrNoRows) {
		return "", false, nil
	}
	return name, err == nil, err
}

// callerRow is the invites routes' head: the caller's users row, or the
// refusal they answer (401 "Invalid user identity" / "User not found").
func (h handlers) callerRow(w http.ResponseWriter, r *http.Request, tx pgx.Tx) (*userRow, bool) {
	caller := policy.UserFrom(r.Context())
	id := parseUUID(&caller.UserID)
	if id == nil {
		refuse(w, http.StatusUnauthorized, "Invalid user identity")
		return nil, false
	}
	user, err := userByID(r.Context(), tx, *id)
	if err != nil {
		h.fail(w, r, "load user", err)
		return nil, false
	}
	if user == nil {
		refuse(w, http.StatusUnauthorized, "User not found")
		return nil, false
	}
	return user, true
}

// joinByInvite validates token, accepts it for user and audits the join. It
// writes the refusal itself and returns false on any failure.
func (h handlers) joinByInvite(w http.ResponseWriter, r *http.Request, tx pgx.Tx, user *userRow, token, description string, now time.Time) (*membershipRow, string, bool) {
	ctx := r.Context()
	invite, err := h.validateInvite(ctx, tx, token, now)
	if err != nil {
		h.fail(w, r, "validate invite", err)
		return nil, "", false
	}
	if invite == nil {
		refuse(w, http.StatusBadRequest, "Invalid or expired invite")
		return nil, "", false
	}
	orgName, found, err := organizationName(ctx, tx, invite.OrgID)
	if err != nil {
		h.fail(w, r, "load organization", err)
		return nil, "", false
	}
	if !found {
		refuse(w, http.StatusNotFound, "Organization not found")
		return nil, "", false
	}
	membershipID, err := h.acceptInvite(ctx, tx, invite, user.ID, now)
	var refused errInviteRefused
	if errors.As(err, &refused) {
		refuse(w, http.StatusBadRequest, refused.message)
		return nil, "", false
	}
	if err != nil {
		h.fail(w, r, "accept invite", err)
		return nil, "", false
	}
	changes := pyjson.NewObject()
	changes.Set("invite_id", invite.ID.String())
	changes.Set("user_id", user.ID.String())
	changes.Set("role", invite.Role)
	userID := user.ID
	if err := h.emitAudit(ctx, tx, r, auditEntry{orgID: invite.OrgID, action: audit.ActionMemberJoined, resourceType: audit.ResourceMembership,
		resourceID: membershipID.String(), userID: &userID, description: description, changes: changes}); err != nil {
		h.fail(w, r, "audit", err)
		return nil, "", false
	}
	role := invite.Role
	joined := now
	return &membershipRow{OrgID: invite.OrgID, Role: &role, JoinedAt: &joined, CreatedAt: now}, orgName, true
}

// commitThenIssue commits tx and then, as _issue_membership_tokens runs
// after the route's commit, mints and stores the token pair in a second
// transaction, and writes the membership token response.
func (h handlers) commitThenIssue(w http.ResponseWriter, r *http.Request, tx pgx.Tx, user *userRow, membership *membershipRow, orgName string) {
	ctx := r.Context()
	if err := tx.Commit(ctx); err != nil {
		h.fail(w, r, "commit", err)
		return
	}
	issue, err := h.Pool.Begin(ctx)
	if err != nil {
		h.fail(w, r, "begin", err)
		return
	}
	defer func() { _ = issue.Rollback(context.Background()) }()
	pair, err := h.issueMembershipTokens(ctx, issue, r, user, membership)
	if err != nil {
		h.fail(w, r, "issue tokens", err)
		return
	}
	if err := issue.Commit(ctx); err != nil {
		h.fail(w, r, "commit tokens", err)
		return
	}
	out := pyjson.NewObject()
	out.Set("access_token", pair.access)
	out.Set("refresh_token", pair.refresh)
	out.Set("token_type", "bearer")
	out.Set("expires_in", accessExpiresIn)
	out.Set("org_id", membership.OrgID.String())
	out.Set("org_name", orgName)
	out.Set("role", roleText(membership.Role))
	policy.WriteModel(w, http.StatusOK, out, nil)
}

// acceptInviteRoute is POST /api/v1/auth/accept-invite (invites.py).
func (h handlers) acceptInviteRoute(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	body, _ := policy.BodyFrom(ctx)
	var errs pybody.Errors
	var token string
	if object, ok := errs.Object(body); ok {
		token, _ = errs.RequiredString(object, "token", 0, 0)
	}
	if len(errs) > 0 {
		writeValidation(w, errs)
		return
	}
	tx, err := h.Pool.Begin(ctx)
	if err != nil {
		h.fail(w, r, "begin", err)
		return
	}
	defer func() { _ = tx.Rollback(context.Background()) }()
	user, ok := h.callerRow(w, r, tx)
	if !ok {
		return
	}
	membership, orgName, ok := h.joinByInvite(w, r, tx, user, token, "Invite accepted", h.Now())
	if !ok {
		return
	}
	h.commitThenIssue(w, r, tx, user, membership, orgName)
}

// validateOrganizationName is services/users.py validate_organization_name.
func validateOrganizationName(name *string) (string, error) {
	normalized := ""
	if name != nil {
		normalized = pythonparity.Strip(*name)
	}
	if normalized == "" {
		return "", errors.New("Workspace name is required")
	}
	if len(pyjson.Runes(normalized)) > 100 {
		return "", errors.New("Workspace name must be 100 characters or fewer")
	}
	return normalized, nil
}

// onboard is POST /api/v1/auth/onboard (invites.py).
func (h handlers) onboard(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	body, _ := policy.BodyFrom(ctx)
	var errs pybody.Errors
	var action string
	var orgName, inviteCode *string
	if object, ok := errs.Object(body); ok {
		action, _ = errs.RequiredString(object, "action", 0, 0)
		if value, present := errs.OptionalString(object, "org_name", 0, 0); present {
			orgName = &value
		}
		if value, present := errs.OptionalString(object, "invite_code", 0, 0); present {
			inviteCode = &value
		}
	}
	if len(errs) > 0 {
		writeValidation(w, errs)
		return
	}
	tx, err := h.Pool.Begin(ctx)
	if err != nil {
		h.fail(w, r, "begin", err)
		return
	}
	defer func() { _ = tx.Rollback(context.Background()) }()
	user, ok := h.callerRow(w, r, tx)
	if !ok {
		return
	}
	var onboarded bool
	if err := tx.QueryRow(ctx, `SELECT EXISTS (SELECT 1 FROM memberships WHERE user_id = $1::uuid)`, user.ID).Scan(&onboarded); err != nil {
		h.fail(w, r, "load memberships", err)
		return
	}
	if onboarded {
		refuse(w, http.StatusBadRequest, "Already onboarded")
		return
	}
	now := h.Now()
	if action == "join_org" {
		if inviteCode == nil || *inviteCode == "" {
			refuse(w, http.StatusBadRequest, "invite_code is required")
			return
		}
		membership, name, ok := h.joinByInvite(w, r, tx, user, *inviteCode, "Invite accepted during onboarding", now)
		if !ok {
			return
		}
		h.commitThenIssue(w, r, tx, user, membership, name)
		return
	}
	if action != "create_org" {
		refuse(w, http.StatusBadRequest, "Invalid action. Use 'create_org' or 'join_org'")
		return
	}
	name, err := validateOrganizationName(orgName)
	if err != nil {
		refuseViolations(w, http.StatusUnprocessableEntity, err.Error(), []string{err.Error()})
		return
	}
	orgID := h.NewUUID()
	if err := insertOrganization(ctx, tx, orgID, slugifyOrgName(name)+"-"+user.ID.String()[:8], name, now); err != nil {
		h.fail(w, r, "insert organization", err)
		return
	}
	if err := insertMembership(ctx, tx, h.NewUUID(), user.ID, orgID, "owner", nil, now); err != nil {
		h.fail(w, r, "insert membership", err)
		return
	}
	changes := pyjson.NewObject()
	changes.Set("organization_name", name)
	userID := user.ID
	if err := h.emitAudit(ctx, tx, r, auditEntry{orgID: orgID, action: audit.ActionCreate, resourceType: audit.ResourceOrganization,
		resourceID: orgID.String(), userID: &userID, description: "Organization created during onboarding", changes: changes}); err != nil {
		h.fail(w, r, "audit", err)
		return
	}
	role := "owner"
	h.commitThenIssue(w, r, tx, user, &membershipRow{OrgID: orgID, Role: &role, JoinedAt: &now, CreatedAt: now}, name)
}

// slugifyOrgName is common._slugify_org_name:
// re.sub(r"[^a-z0-9]+", "-", name.lower()).strip("-")[:50], or
// "my-organization" when that is empty. The class is ASCII, so every run of
// any other code point, "-" included, becomes one "-".
func slugifyOrgName(name string) string {
	var slug []rune
	inRun := false
	for _, r := range pyjson.Runes(pythonparity.Lower(name)) {
		if (r >= 'a' && r <= 'z') || (r >= '0' && r <= '9') {
			slug = append(slug, r)
			inRun = false
			continue
		}
		if !inRun {
			slug = append(slug, '-')
			inRun = true
		}
	}
	for len(slug) > 0 && slug[0] == '-' {
		slug = slug[1:]
	}
	for len(slug) > 0 && slug[len(slug)-1] == '-' {
		slug = slug[:len(slug)-1]
	}
	if len(slug) > 50 {
		slug = slug[:50]
	}
	if len(slug) == 0 {
		return "my-organization"
	}
	return string(slug)
}
